### ISSUE TITLE
ITicketStore from DI

### ISSUE THREAD

**Issue by leastprivilege** [aspnet/AspNetCore#6946]
Wouldn't it make sense to resolve the ITicketStore for the cookie handler from DI in the future?

**Comment by PinpointTownes** [comment]
FYI, it's been discussed here: https://github.com/aspnet/Security/pull/1719.

**Comment by leastprivilege** [comment]
I am glad you have the time to search the archives ;)

**Comment by Eilon** [comment]
Marking as up-for-grabs in case someone wants to take a crack at this. We don't currently plan to implement this ourselves, but we'd be open to a collaboration/contribution for this.

**Comment by crozone** [comment]
We're currently implementing a `DbContext` based implementation of `ITicketStore`, and this requires `ITicketStore` to be accessible via DI in a `Scoped` context.

The solution should be very similar to the implementation for `services.AddDataProtection().PersistKeysToDbContext<ApplicationDbContext>();`, however the `AddCookie()` extension provides no access to the DI container.

The current solution is to use `services.AddOptions<CookieAuthenticationOptions>(CookieAuthenticationDefaults.AuthenticationScheme).Configure<ITicketStore>((o, t) => o.SessionStore = t);`, but we cannot make guarantees about `DbContext` lifetime. If we discover a cleaner way to do this, I'll look into submitting a PR.

**Comment by niklasholm** [comment]
Also, it would be neat to have some default implementation(s) for `ITicketStore` available, for example one using `ISession` to store the data.

**Comment by Tratcher** [comment]
FYI this is much improved by https://github.com/dotnet/aspnetcore/pull/19268.
`.AddCookie<MemoryCacheTicketStore>((o, ticketStore) => o.SessionStore = ticketStore);`

**Comment by kevinchalet** [comment]
@Tratcher it only works with singleton services, right?

**Comment by Tratcher** [comment]
Correct, but a singleton can inject the IServiceProvider and create its own scope.

**Comment by kevinchalet** [comment]
Interesting. Why not adding a `SessionStoreType` property like what we did for `EventsType`? It works amazingly well.

**Comment by Tratcher** [comment]
We can, but it's never become a priority. The new extensions are a general feature for a general problem.

**Comment by AnthonyMastrean** [comment]
> FYI this is much improved by #19268.
> `.AddCookie<MemoryCacheTicketStore>((o, ticketStore) => o.SessionStore = ticketStore);`

These are greatly improved signatures... but don't these services get resolved from the root service provider? Doesn't that mean we can't resolve scoped dependencies in the `ITicketStore` implementation type? There doesn't seem to be any way to use the request scope or to create short-lived scopes... is our only option to inject a scope factory and create a local scope that lives for the lifetime of the store?

**Comment by juantxorena** [comment]
6 years later, what is the proper way of using ITicketStore from DI? A method was implemented in #19268 as mentioned, but it was reverted in #24253 with no alternative.
